src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs

```diff
@@ -3,38 +3,33 @@
 using System;
 using System.Text;
 using System.Net.Security;
+using System.Buffers;
 
 #nullable enable
 
 namespace Microsoft.Data.SqlClient
 {
     internal sealed class NegotiateSSPIContextProvider : SSPIContextProvider
     {
-        private NegotiateAuthentication? _negotiateAuth = null;
-
-        internal override void GenerateSspiClientContext(ReadOnlyMemory<byte> received, ref byte[] sendBuff, ref uint sendLength, byte[][] _sniSpnBuffer)
+        protected override void GenerateSspiClientContext(ReadOnlySpan<byte> incomingBlob, IBufferWriter<byte> outgoingBlobWriter, byte[][] _sniSpnBuffer)
         {
-            NegotiateAuthenticationStatusCode statusCode = NegotiateAuthenticationStatusCode.UnknownCredentials;
+            NegotiateAuthenticationStatusCode statusCode = default;
 
             for (int i = 0; i < _sniSpnBuffer.Length; i++)
             {
-                _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = Encoding.Unicode.GetString(_sniSpnBuffer[i]) });
-                sendBuff = _negotiateAuth.GetOutgoingBlob(received.Span, out statusCode)!;
+                var negotiateAuth = new NegotiateAuthentication(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = Encoding.Unicode.GetString(_sniSpnBuffer[i]) });
+                var result = negotiateAuth.GetOutgoingBlob(incomingBlob, out statusCode);
+
                 // Log session id, status code and the actual SPN used in the negotiation
                 SqlClientEventSource.Log.TryTraceEvent("{0}.{1} | Info | Session Id {2}, StatusCode={3}, SPN={4}", nameof(NegotiateSSPIContextProvider),
-                    nameof(GenerateSspiClientContext), _physicalStateObj.SessionId, statusCode, _negotiateAuth.TargetName);
+                    nameof(GenerateSspiClientContext), _physicalStateObj.SessionId, statusCode, negotiateAuth.TargetName);
                 if (statusCode == NegotiateAuthenticationStatusCode.Completed || statusCode == NegotiateAuthenticationStatusCode.ContinueNeeded)
-                    break; // Successful case, exit the loop with current SPN.
-                else
-                    _negotiateAuth = null; // Reset _negotiateAuth to be generated again for next SPN.
-            }
-
-            if (statusCode is not NegotiateAuthenticationStatusCode.Completed and not NegotiateAuthenticationStatusCode.ContinueNeeded)
-            {
-                throw new InvalidOperationException(SQLMessage.SSPIGenerateError() + Environment.NewLine + statusCode);
+                {
+                    outgoingBlobWriter.Write(result);
```

Comment:
In a recent fix in main (now merged in here), the for loop was added to try authenticating with each SPN provided in the _sniSpnBuffer array (as opposed to just the first one).
Are you missing a break to exit the loop on success of the current SPN here? (I think this is causing some of the CI failures.)
```suggestion
                    outgoingBlobWriter.Write(result);
                    break;
```